캠핑용품 대여 및 반납 후 종료 기능

Habist opened the pull request:
## 작업 내용
제공자가 사용자에게 캠핑용품을 대여하고 상태를 대여 중으로 변경하는 기능을 추가했습니다.
사용자는 캠핑용품 대여가 끝난 후 반납하고 상태를 거래 종료로 변경하는 기능을 추가하였습니다.

## 주의 사항
commit 이력이 충돌하여 이전에 merge한 내용이 보여지고 있습니다. -> conflict resolve 완료
Files changed는 변경 사항만 포함됩니다.

>  참조이슈
> #53

Habist committed:
refactor: [#64] Mapper 인터페이스 메소드 명 수정

select -> find 일괄 변경하였습니다.
insert -> insertXXX 구체적으로 어떤 모델을 넣는지 명시하였습니다.

Habist committed:
refactor: [#64] 쿼리 문자 일관성 있게 변경

리눅스 환경에선 대소문자를 구분하기 때문에 일관성 있도록 테이블 문자를 대문자로 변경하였습니다.
컬럼 또한 위와 같습니다.

Habist committed:
refactor: [#64] 예약 정보 요청 쿼리 Slave 노드로 분기 처리

Transactional 애노테이션의 readOnly 속성으로 Slave 노드로 분기 처리하도록 추가하였습니다.

Habist committed:
refactor: [#64] 렌탈 데이터 상태 Enum 클래스로 관리하도록 변경

렌탈 데이터의 상태를 관리하기 위해 렌탈 상태를 나타내는 Enum 클래스 추가하였습니다.

Habist committed:
refactor: [#64] 예약 생성 시 렌탈 데이터 상태 변경 로직 추가

예약 생성 시 렌탈 데이터 상태를 거래중으로 변경하도록 추가하였습니다.

Habist committed:
feat: [#53] 캠핑용품 대여 완료 기능 추가

예약된 상태에서 제공자가 캠핑용품을 전달 후 대여 완료 표시를 위한 기능을 추가하였습니다.

Habist committed:
feat: [#53] 캠핑용품 대여 완료 시 요청자 아이디 검증 로직 추가

대여 완료 요청이 제공자의 아이디가 맞는지 확인하는 로직을 추가하였습니다.

Habist committed:
refactor: [#53] 대여, 예약 model 클래스 status 필드 타입 변경

status 필드의 타입을 좀 더 명확하고 안전한 Enum 타입으로 변경하였습니다.

Habist committed:
refactor: [#53] Status 비교 Enum 클래스 기반으로 변경

Habist committed:
feat: [#53] 대여 완료 및 거래 완료 기능 추가

Habist committed:
Merge branch 'develop' into feature/53

f-lab-dev commented on src/main/java/com/campool/controller/RentalController.java:
```diff
@@ -48,4 +56,18 @@ public RentalDetailsResponse getRentalInfoById(@PathVariable long id) {
     }
 
+    @LoginValidation
+    @PatchMapping("/rentals/rent/{rentalId}")
```
여기는 `/rentals/rent/` 이렇게 구성하신 이유가 있으실까요?

Habist replied:
처음에는 `PATCH` `/rentals/{rentalId}`를 고민했었는데 이렇게 구성하면 API가 **대여(rent)** 를 위한 요청보다는 rentals 리소스 자체의 변경을 뜻하는 것 같아 애매한 의미가 될 수 있을 것이라 생각했습니다. 

그래서 **대여(rent)** 라는 의미를 명확하게 하기 위해서 `rent`를 추가하여 구성하였습니다.

f-lab-dev replied:
그럼 그대로 읽었을 때 의미가 잘못 전달 될 수 있을 것 같은데 `/rentals/{rentalId}/rent`는 어떠실까요?

Habist replied:
좋은 것 같습니다~ 리뷰 반영할게요!

f-lab-dev commented on src/main/java/com/campool/service/RentalService.java:
```diff
@@ -74,4 +79,30 @@ private String getPolygonString(BigDecimal longitude, BigDecimal latitude, BigDe
+    }
+
+    private boolean isValidRentalStatus(RentalInfo rentalInfo, RentalStatus status, String userId) {
+        return rentalInfo.getStatus().equals(status) && rentalInfo.getUserId().equals(userId);
```
status는 `equals`를 이용해 비교할 필요가 있을까요?

Habist replied:
> [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se8/jls8.pdf)
> 8.9 Enum
> it is permitted to use the
> == operator in place of the equals method when comparing two object references
> if it is known that at least one of them refers to an enum constant.

위와 같이 자바 스펙에 명시되어 있고 `==`을 사용하면 이점도 많은 것 같습니다.
- `NullPointerException` 발생하지 않음
- 컴파일 타임에 다른 타입과 비교하면 예외 발생하여 훨씬 타입 안전함

 `==` 비교로 수정하여 반영하겠습니다!
피드백 감사합니다~

f-lab-dev commented on src/main/java/com/campool/service/RentalService.java:
```diff
@@ -74,4 +79,30 @@ private String getPolygonString(BigDecimal longitude, BigDecimal latitude, BigDe
+    @Transactional
+    public void updateNewStatus(RentalStatus currentStatus, RentalStatus newStatus, long rentalId,
+            String userId) {
+        if (isValidRentalStatus(getRentalById(rentalId), currentStatus, userId)) {
+            rentalMapper.updateStatusById(rentalId, newStatus);
+        } else {
+            throw new IllegalStateException(currentStatus.getMessage() + "인 상태가 아닙니다.");
+        }
```
조금 더 간결하게 표현이 가능할 것 같은데 어떠실까요?

```suggestion
        if (!isValidRentalStatus(getRentalById(rentalId), currentStatus, userId)) {
            throw new IllegalStateException(currentStatus.getMessage() + "인 상태가 아닙니다.");
        }
        
        rentalMapper.updateStatusById(rentalId, newStatus);
```

Habist replied:
`논리 부정 연산자`를 사용하면 막연히 가독성이 떨어질 것 같아서 사용하지 않았는데 **간결한 표현**을 생각하면 사용하는 것이 좋다는 생각이 드네요! 

리뷰 반영하겠습니다! 
감사합니다~

Habist committed:
refactor: [#53 - 리뷰 반영] Enum 비교 로직 equals -> == 로 변경

Habist committed:
refactor: [#53] 사용하지 않는 bookingMapper 의존성 제거

Habist committed:
refactor: [#53 - 리뷰 반영] 논리 부정 연산자를 사용하여 코드 간결하게 표현

Habist committed:
refactor: [#53 - 리뷰 반영] 논리 부정 연산자를 사용하여 코드 간결하게 표현

Habist committed:
Merge branch 'feature/53' of https://github.com/f-lab-edu/campool into feature/53

Habist committed:
refactor: [#53 - 리뷰 반영] Enum 클래스 비교 추가 수정

Habist committed:
refactor: [#53 - 리뷰 반영] 캠핑용품 대여 요청 url 변경

Habist committed:
예약 및 결제 취소 기능 (#71)

* feat: [#68] 예약 및 결제 취소 기능 추가

대여 전 상태의 예약을 취소할 수 있는 기능을 추가하였습니다.
iamport 클라이언트 생성 부분이 공통적으로 나타나 메소드로 추출하였습니다.

* refactor: [#68 - 리뷰반영] 가독성을 위해 유효성 검증 부분 메소드로 표현

* feat: [#69] 대여 등록한 캠핑용품 삭제 기능 추가 (#73)
